railties/lib/rails/rack/logger.rb

```diff
@@ -30,11 +30,14 @@ def call(env)
       private
         def call_app(request, env) # :doc:
           instrumenter = ActiveSupport::Notifications.instrumenter
-          handle = instrumenter.build_handle("request.action_dispatch", { request: request })
+          payload = { request: request }
+          handle = instrumenter.build_handle("request.action_dispatch", payload)
           handle.start
 
           logger.info { started_request_message(request) }
           status, headers, body = response = @app.call(env)
+          payload[:status] = status
+          payload[:headers] = headers
```

Comment:
I wonder if this should `dup` & `freeze` the hash so that a notifier can't modify the response - I assume it's not _supposed_ to be able to.

If that's a performance concern, we could instead create a kind of proxy object which raised on `[]=` and friends?

Reply:
I'm very confused about how this is supposed to work.  The downstream API (whoever listens for the `start` event) must keep a reference to the payload and then hope it gets mutated later when the request finishes?

Is there a reason we can't pass the status / headers to the finish method?  I think there's a [finish with values](https://github.com/skipkayhil/rails/blob/d21d811ffece4d3959bcd37e58fec77590ff6f93/activesupport/lib/active_support/notifications/fanout.rb#L227) method, just not sure if it applies here.

Additionally, it seems like it's possible for `@app.call` to raise an exception. I assume the whoever is interested in these events would want to know that?

Reply:
> The downstream API (whoever listens for the start event) must keep a reference to the payload and then hope it gets mutated later when the request finishes?

The instrumenter itself actually saves the payload: https://github.com/rails/rails/blob/c32813ddc6e7e46a924bf98574d1855b494967e8/activesupport/lib/active_support/notifications/instrumenter.rb#L46. So if you just listen for the event with a subscriber, your event payload will include these fields.

> Is there a reason we can't pass the status / headers to the finish method? I think there's a finish with values method, just not sure if it applies here.

We could rewrite this to use `finish_with_values`, but it would behave the same from the perspective of code listening to the event (although code using an evented subscriber could in theory notice the payload from the start/finish phases has a different `object_id`

> Additionally, it seems like it's possible for @app.call to raise an exception. I assume the whoever is interested in these events would want to know that?

Agreed, there's a rescue block further down which is finishing the notification, we could include the exception in the payload there.

That said, it "shouldn't" happen; this middleware goes below `ActionDispatch::DebugExceptions`, so any exceptions here have come from Rails and are going to cause the web server to return some kind of default 500 error

Reply:
Are you happy with mutating the payload given that it's saved by the notifier, or would you like me (or @skipkayhil!) to open a new PR with `#finish_with_values`?

Reply:
Poke @skipkayhil - is there anything I can do to help get this moving? Would you like me to open another PR which deals with the exception in the rescue block?